File: src/frontend/src/lib/api/invite.ts
import {
  mapAcceptOrgInviteRequest,
  mapAcceptOrgInviteResponse,
  mapCreateOrgInviteRequest,
  mapCreateOrgInviteResponse,
  mapDeclineOrgInviteRequest,
  mapDeclineOrgInviteResponse,
  mapListMyInvitesResponse,
  mapListOrgInvitesRequest,
  mapListOrgInvitesResponse,
  mapRevokeOrgInviteRequest,
  mapRevokeOrgInviteResponse,
  type AcceptOrgInviteRequest,
  type CreateOrgInviteRequest,
  type DeclineOrgInviteRequest,
  type ListOrgInvitesRequest,
  type OrgInvite,
  type RevokeOrgInviteRequest,
} from '@/lib/api-models';
import type { ActorSubclass } from '@icp-sdk/core/agent';
import type { _SERVICE } from '@ssn/backend-api';

export class InviteApi {
  constructor(private readonly actor: ActorSubclass<_SERVICE>) {}

  public async createOrgInvite(
    req: CreateOrgInviteRequest,
  ): Promise<OrgInvite> {
    const res = await this.actor.create_org_invite(
      mapCreateOrgInviteRequest(req),
    );
    return mapCreateOrgInviteResponse(res);
  }

  public async listOrgInvites(
    req: ListOrgInvitesRequest,
  ): Promise<OrgInvite[]> {
    const res = await this.actor.list_org_invites(
      mapListOrgInvitesRequest(req),
    );
    return mapListOrgInvitesResponse(res);
  }

  public async revokeOrgInvite(req: RevokeOrgInviteRequest): Promise<void> {
    const res = await this.actor.revoke_org_invite(
      mapRevokeOrgInviteRequest(req),
    );
    mapRevokeOrgInviteResponse(res);
  }

  public async listMyInvites(): Promise<OrgInvite[]> {
    const res = await this.actor.list_my_invites();
    return mapListMyInvitesResponse(res);
  }

  public async acceptOrgInvite(req: AcceptOrgInviteRequest): Promise<void> {
    const res = await this.actor.accept_org_invite(
      mapAcceptOrgInviteRequest(req),
    );
    mapAcceptOrgInviteResponse(res);
  }

  public async declineOrgInvite(req: DeclineOrgInviteRequest): Promise<void> {
    const res = await this.actor.decline_org_invite(
      mapDeclineOrgInviteRequest(req),
    );
    mapDeclineOrgInviteResponse(res);
  }
}

File: src/backend/src/controller/invite.rs
use crate::{
    dto::{
        AcceptOrgInviteRequest, AcceptOrgInviteResponse, CreateOrgInviteRequest,
        CreateOrgInviteResponse, DeclineOrgInviteRequest, DeclineOrgInviteResponse,
        ListMyInvitesResponse, ListOrgInvitesRequest, ListOrgInvitesResponse,
        RevokeOrgInviteRequest, RevokeOrgInviteResponse,
    },
    service::invite_service,
};
use canister_utils::{assert_authenticated, ApiResultDto};
use ic_cdk::{
    api::{msg_caller, time},
    *,
};

#[update]
fn create_org_invite(req: CreateOrgInviteRequest) -> ApiResultDto<CreateOrgInviteResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    invite_service::create_org_invite(&caller, req, time()).into()
}

#[query]
fn list_org_invites(req: ListOrgInvitesRequest) -> ApiResultDto<ListOrgInvitesResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    invite_service::list_org_invites(&caller, req, time()).into()
}

#[update]
fn revoke_org_invite(req: RevokeOrgInviteRequest) -> ApiResultDto<RevokeOrgInviteResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    invite_service::revoke_org_invite(&caller, req).into()
}

#[query]
fn list_my_invites() -> ApiResultDto<ListMyInvitesResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    invite_service::list_my_invites(&caller, time()).into()
}

#[update]
fn accept_org_invite(req: AcceptOrgInviteRequest) -> ApiResultDto<AcceptOrgInviteResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    invite_service::accept_org_invite(&caller, req, time()).into()
}

#[update]
fn decline_org_invite(req: DeclineOrgInviteRequest) -> ApiResultDto<DeclineOrgInviteResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    invite_service::decline_org_invite(&caller, req, time()).into()
}
